pkg/traceql/storage.go

```diff
@@ -85,9 +85,9 @@ type Span interface {
 	StartTimeUnixNanos() uint64
 	DurationNanos() uint64
 
-	SiblingOf(Span) bool
-	DescendantOf(Span) bool
-	ChildOf(Span) bool
+	SiblingOf([]Span, []Span, bool, bool, []Span) []Span
```

Comment:
Do we really invert or could the engine swap lhs and rhs? I think falseForAll is a required though.

Thinking out loud on other ways to organize this:

1) Put them on the Fetcher?
2) Put them on the Spanset: `lhs.SiblingsOf(rhs, bool, bool, buffer)`. But looking at that, doesn't really seem much better.   And we'd need to make Spanset an interface.
3) Expose left/right/etc to the engine move the algorithm there.

I think only your current interface and Fetcher are practical for now.  Is there any drawback to putting it on the Fetcher?

Reply:
> Do we really invert or could the engine swap lhs and rhs? I think falseForAll is a required though.
 
 Unfortunately, I think we do b/c we always return the right hand side. invert just changes the direction that the relationship is evaluated. Child becomes parent. We could make 3 more functions (the inversions) and that would allow us to drop it

> Put them on the Fetcher?

My only pushback on this suggestion is that we'd have to change the pipelineElement interface to pass the fetcher down through all of the elements. It would be a rather sweeping (although straighforward) change.

```
type pipelineElement interface {
 	evaluate([]*Spanset, *SpansetFetcher) ([]*Spanset, error)
}
```

I like this suggestion about the same as I like keeping it where it is. Which is to say I don't think either option is great, but both work and are practical. Both options muddy the waters and make writing new fetch layers for the engine a bit more unwieldy.